src/Sulu/Bundle/CategoryBundle/Controller/CategoryController.php

```diff
@@ -353,27 +305,73 @@ protected function getCategoryListRepresentation(Request $request, $parentKey =
+     * for the given locale. The returned ListBuilder is initialized with the request-parameters and respective
+     * select fields.
+     *
+     * @param $locale
      *
-     * @param $parentKey
-     * @param DoctrineListBuilder $listBuilder
+     * @return DoctrineListBuilder
      */
-    protected function addParentSelector($parentKey, DoctrineListBuilder $listBuilder)
+    private function initializeListBuilder($locale)
     {
-        $manager = $this->getManager();
-        $parentEntity = $manager->findByKey($parentKey);
-
-        $listBuilder->between(
-            new DoctrineFieldDescriptor(
-                'lft',
-                'lft',
-                CategoryManager::$categoryEntityName,
-                'public.lft',
-                [],
-                true
-            ),
-            [$parentEntity->getLft() + 1, $parentEntity->getRgt()]
-        );
+        /** @var RestHelperInterface $restHelper */
+        $restHelper = $this->get('sulu_core.doctrine_rest_helper');
+
+        /** @var DoctrineListBuilderFactory $factory */
+        $factory = $this->get('sulu_core.doctrine_list_builder_factory');
+
+        $fieldDescriptors = $this->getCategoryManager()->getFieldDescriptors($locale);
+        $listBuilder = $factory->create(self::$entityName);
+        $restHelper->initializeListBuilder($listBuilder, $fieldDescriptors);
+
+        $listBuilder->addSelectField($fieldDescriptors['depth']);
+        $listBuilder->addSelectField($fieldDescriptors['parent']);
+        $listBuilder->addSelectField($fieldDescriptors['locale']);
+        $listBuilder->addSelectField($fieldDescriptors['defaultLocale']);
+        $listBuilder->addSelectField($fieldDescriptors['lft']);
+        $listBuilder->addSelectField($fieldDescriptors['rgt']);
+
+        return $listBuilder;
+    }
+
+    /**
+     * Expands all paths from the given root-level the the given leafs.
+     * All categories which are children of a category along this path are returned.
+     * If a given leaf is not a descendant of the given root-level, it is ignored.
+     *
+     * @param $leafIds array Ids of the categories of the leaves
+     * @param $rootIds array Ids of the categories of the root level
+     * @param $locale
+     *
+     * @return array
+     */
+    private function expandChildren($leafIds, $rootIds, $locale)
+    {
+        // collect ids of categories between root level (inclusive) and leaves (exclusive)
+        $pathCategoryIds = [];
+        foreach ($this->getCategoryManager()->findBetween($rootIds, $leafIds, $locale) as $betweenCategory) {
+            $pathCategoryIds[] = $betweenCategory->getId();
+
+            // add category of root level to path category, if it is connected with a found path
+            if (in_array($betweenCategory->getParent()->getId(), $rootIds)) {
+                $pathCategoryIds[] = $betweenCategory->getParent()->getId();
```

Comment:
Still not very clean, and I am not sure if every database has no problem with that. Can't this loop be rewritten to function correctly?